File: dhttemp.py
#!/usr/bin/env python3
"""
DHT11 Temperature & Humidity Sensor Control Script
Runs on Raspberry Pi - Only handles sensor on/off control
"""

import time
import json
import requests
import threading
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
import board
import adafruit_dht

# ============================================
# Configuration
# ============================================
BACKEND_URL = "https://web-production-3d9a.up.railway.app"  # Your Railway backend
DEVICE_ID = "3d49c55d-bbfd-4bd0-9663-8728d64743ac"  # Raspberry Pi device ID from admin portal
SENSOR_ID = 6  # DHT11 Sensor ID (integer)
DHT_PIN = board.D4  # GPIO4
STATUS_CHECK_INTERVAL = 5  # Check backend status every 5 seconds

def get_local_ip():
    """Get the local IP address of this Raspberry Pi"""
    import socket
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        s.connect(("8.8.8.8", 80))
        ip = s.getsockname()[0]
        s.close()
        return ip
    except Exception:
        return None

def register_device_ip():
    """Register this device's IP address with the backend"""
    try:
        ip_address = get_local_ip()
        if not ip_address:
            print("⚠️  Could not determine local IP address")
            return
        
        print(f"📍 Local IP: {ip_address}")
        
        url = f"{BACKEND_URL}/api/devices/{DEVICE_ID}/metadata"
        response = requests.put(
            url,
            json={"ip_address": ip_address},
            headers={"Content-Type": "application/json"},
            timeout=10
        )
        
        if response.status_code in [200, 201]:
            print(f"✅ Device IP registered: {ip_address}")
        else:
            print(f"⚠️  Failed to register IP: {response.status_code}")
    except Exception as e:
        print(f"⚠️  Error registering IP: {e}")

# ============================================
# Global State
# ============================================
sensor_enabled = True
dht = None

# ============================================
# HTTP Request Handler
# ============================================
class SensorHandler(BaseHTTPRequestHandler):
    """HTTP handler to receive commands from backend"""
    
    def do_GET(self):
        """Handle GET requests"""
        global sensor_enabled
        
        parsed_url = urlparse(self.path)
        path = parsed_url.path
        query = parse_qs(parsed_url.query)
        
        if path == '/sensor/status':
            response = {
                'status': 'ok',
                'enabled': sensor_enabled,
                'device_id': DEVICE_ID,
                'sensor_id': SENSOR_ID,
                'timestamp': time.time()
            }
            self._send_response(200, response)
        
        elif path == '/sensor/control':
            action = query.get('action', [''])[0]
            
            if action == 'on':
                sensor_enabled = True
                print("✅ Sensor turned ON")
                response = {'status': 'Sensor turned ON', 'enabled': True}
                self._send_response(200, response)
            
            elif action == 'off':
                sensor_enabled = False
                print("⏸️  Sensor turned OFF")
                response = {'status': 'Sensor turned OFF', 'enabled': False}
                self._send_response(200, response)
            
            else:
                self._send_response(400, {'error': 'Invalid action. Use ?action=on or ?action=off'})
        
        elif path == '/health':
            self._send_response(200, {'status': 'ok', 'sensor_enabled': sensor_enabled})
        
        else:
            self._send_response(404, {'error': 'Not found'})
    
    def _send_response(self, status_code, data):
        """Send JSON response"""
        self.send_response(status_code)
        self.send_header('Content-type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(json.dumps(data).encode())
    
    def log_message(self, format, *args):
        """Suppress default logging"""
        pass

# ============================================
# Check Sensor Status from Backend
# ============================================
def check_sensor_status():
    """Check if sensor is enabled in backend database"""
    global sensor_enabled
    
    try:
        response = requests.get(
            f"{BACKEND_URL}/api/sensors?deviceId={DEVICE_ID}",
            timeout=5
        )
        
        if response.status_code == 200:
            sensors = response.json()
            for sensor in sensors:
                if sensor.get('sensor_id') == SENSOR_ID:
                    backend_enabled = sensor.get('enabled', True)
                    
                    # Update local state if changed
                    if backend_enabled != sensor_enabled:
                        sensor_enabled = backend_enabled
                        status = "ON" if sensor_enabled else "OFF"
                        print(f"🔄 Sensor state updated from backend: {status}")
                    
                    return backend_enabled
        
        return sensor_enabled
    except Exception as e:
        print(f"⚠️  Error checking backend status: {e}")
        return sensor_enabled

# ============================================
# Status Monitoring Loop
# ============================================
def status_monitor_loop():
    """Monitor sensor status from backend"""
    print(f"🔍 Starting status monitor (Device: {DEVICE_ID}, Sensor: {SENSOR_ID})")
    
    while True:
        try:
            check_sensor_status()
            time.sleep(STATUS_CHECK_INTERVAL)
        
        except KeyboardInterrupt:
            print("🛑 Status monitor interrupted")
            break
        except Exception as e:
            print(f"❌ Error in status monitor: {e}")
            time.sleep(STATUS_CHECK_INTERVAL)

# ============================================
# Start HTTP Server
# ============================================
def start_http_server(port=5000):
    """Start HTTP server to handle commands"""
    server = HTTPServer(('0.0.0.0', port), SensorHandler)
    print(f"🌐 HTTP Server started on port {port}")
    print(f"📍 Control Endpoints:")
    print(f"   - GET http://localhost:{port}/sensor/status")
    print(f"   - GET http://localhost:{port}/sensor/control?action=on")
    print(f"   - GET http://localhost:{port}/sensor/control?action=off")
    print(f"   - GET http://localhost:{port}/health")
    
    server.serve_forever()

# ============================================
# Main Entry Point
# ============================================
if __name__ == "__main__":
    try:
        print("🚀 Initializing DHT11 Sensor Control...")
        print("ℹ️  This script only handles sensor on/off control")
        print("ℹ️  No sensor data will be sent to backend")
        
        # Register device IP with backend
        register_device_ip()
        
        # Start HTTP server in background thread
        server_thread = threading.Thread(target=start_http_server, args=(5000,), daemon=True)
        server_thread.start()
        
        # Run status monitor in main thread
        status_monitor_loop()
    
    except Exception as e:
        print(f"❌ Fatal error: {e}")
    
    finally:
        print("🔌 Cleaning up...")
        if dht:
            dht.exit()
        print("✅ Goodbye!")

File: dhttemp.js
#!/usr/bin/env node
/**
 * DHT11 Sensor Control Agent (Raspberry Pi)
 *
 * Control-only behavior:
 * - Registers a controllable sensor row in EC2 PostgreSQL via Sensor Control API.
 * - Polls desired on/off state from EC2 API.
 * - Does NOT send sensor readings/data to backend.
 */

const express = require('express');
const os = require('os');
const fs = require('fs');

require('dotenv').config();

// ============================================
// Configuration
// ============================================
const SENSOR_CONTROL_API_URL = (process.env.SENSOR_CONTROL_API_URL || 'http://13.205.201.82/sensor-api').replace(/\/$/, '');
const DEVICE_ID_FILE = process.env.DEVICE_ID_FILE || './device_id.txt';
const DEVICE_ID = (process.env.DEVICE_ID || (fs.existsSync(DEVICE_ID_FILE) ? fs.readFileSync(DEVICE_ID_FILE, 'utf8').trim() : '')).trim();
const SENSOR_NAME = process.env.SENSOR_NAME || 'DHT11 Sensor';
const SENSOR_TYPE = process.env.SENSOR_TYPE || 'dht11';
const API_KEY = process.env.API_KEY || process.env.ADMIN_API_KEY || '';
const DEVICE_SECRET = process.env.DEVICE_REGISTRATION_SECRET || '';
const STATUS_CHECK_INTERVAL = Number(process.env.SENSOR_STATUS_CHECK_INTERVAL_MS || 5000);
const HTTP_PORT = Number(process.env.SENSOR_LOCAL_PORT || 5000);

// ============================================
// Global State
// ============================================
let sensorEnabled = true;
let sensorId = null;

if (!DEVICE_ID) {
  console.error('❌ DEVICE_ID is required (set DEVICE_ID or DEVICE_ID_FILE)');
  process.exit(1);
}

if (!API_KEY) {
  console.error('❌ API_KEY (or ADMIN_API_KEY) is required for sensor-control API calls');
  process.exit(1);
}

// ============================================
// Utility Functions
// ============================================

async function requestJson(path, { method = 'GET', body, extraHeaders = {} } = {}) {
  const response = await fetch(`${SENSOR_CONTROL_API_URL}${path}`, {
    method,
    headers: {
      'Content-Type': 'application/json',
      'x-api-key': API_KEY,
      ...extraHeaders,
    },
    body: body ? JSON.stringify(body) : undefined,
  });

  const text = await response.text();
  let json;
  try {
    json = text ? JSON.parse(text) : {};
  } catch {
    json = { raw: text };
  }

  if (!response.ok) {
    throw new Error(`HTTP ${response.status}: ${json.error || json.message || text || 'Request failed'}`);
  }

  return json;
}

/**
 * Get local IP for diagnostics
 */
function getLocalIP() {
  try {
    const interfaces = os.networkInterfaces();
    for (const name of Object.keys(interfaces)) {
      for (const iface of interfaces[name]) {
        // Skip internal and non-IPv4 addresses
        if (iface.family === 'IPv4' && !iface.internal) {
          return iface.address;
        }
      }
    }
    return null;
  } catch (error) {
    console.error('Error getting local IP:', error);
    return null;
  }
}

/**
 * Register or upsert this Pi's DHT sensor on EC2
 */
async function registerSensor() {
  try {
    const response = await requestJson('/api/sensors/register', {
      method: 'POST',
      body: {
        deviceId: DEVICE_ID,
        sensorName: SENSOR_NAME,
        sensorType: SENSOR_TYPE,
      },
      extraHeaders: {
        'x-device-secret': DEVICE_SECRET,
      },
    });

    sensorId = response?.sensor?.sensor_id || null;
    if (sensorId) {
      console.log(`✅ Sensor registered on EC2 (sensor_id=${sensorId})`);
    } else {
      console.warn('⚠️  Sensor registration succeeded but sensor_id missing in response');
    }
  } catch (error) {
    console.error(`❌ Failed to register sensor on EC2: ${error.message}`);
    throw error;
  }
}

/**
 * Poll desired sensor state from EC2 backend
 */
async function checkSensorStatus() {
  try {
    const sensors = await requestJson(`/api/sensors?deviceId=${encodeURIComponent(DEVICE_ID)}`);
    const target = Array.isArray(sensors)
      ? sensors.find((s) => (sensorId ? s.sensor_id === sensorId : String(s.sensor_name || '').toLowerCase() === SENSOR_NAME.toLowerCase())) || sensors[0]
      : null;

    if (!target) {
      console.warn('⚠️  No sensor row found for device yet');
      return sensorEnabled;
    }

    if (!sensorId && target.sensor_id) {
      sensorId = target.sensor_id;
    }

    const backendEnabled = target.enabled !== false;
    if (backendEnabled !== sensorEnabled) {
      sensorEnabled = backendEnabled;
      console.log(`🔄 Sensor state changed from EC2 backend: ${sensorEnabled ? 'ON' : 'OFF'}`);
    }

    return sensorEnabled;
  } catch (error) {
    console.warn(`⚠️  Error polling sensor status: ${error.message}`);
    return sensorEnabled;
  }
}

/**
 * Status monitoring loop - polls desired state from EC2
 */
async function statusMonitorLoop() {
  console.log(`🔍 Starting status monitor (device=${DEVICE_ID}, interval=${STATUS_CHECK_INTERVAL}ms)`);

  while (true) {
    try {
      await checkSensorStatus();
      await new Promise((resolve) => setTimeout(resolve, STATUS_CHECK_INTERVAL));
    } catch (error) {
      console.error(`❌ Error in status monitor: ${error.message}`);
      await new Promise((resolve) => setTimeout(resolve, STATUS_CHECK_INTERVAL));
    }
  }
}

// ============================================
// Local Express Server (Optional local control/debug)
// ============================================

const app = express();
app.use(express.json());

app.get('/sensor/status', (req, res) => {
  res.json({
    status: 'ok',
    enabled: sensorEnabled,
    device_id: DEVICE_ID,
    sensor_id: sensorId,
    timestamp: Date.now() / 1000,
  });
});

app.get('/sensor/control', (req, res) => {
  const action = String(req.query.action || '').toLowerCase();

  if (action === 'on') {
    sensorEnabled = true;
    console.log('✅ Sensor turned ON (local endpoint)');
    return res.json({ status: 'Sensor turned ON', enabled: true });
  }

  if (action === 'off') {
    sensorEnabled = false;
    console.log('⏸️  Sensor turned OFF (local endpoint)');
    return res.json({ status: 'Sensor turned OFF', enabled: false });
  }

  return res.status(400).json({ error: 'Invalid action. Use ?action=on or ?action=off' });
});

app.get('/health', (req, res) => {
  res.json({ status: 'ok', sensor_enabled: sensorEnabled, sensor_id: sensorId });
});

app.use((req, res) => {
  res.status(404).json({ error: 'Not found' });
});

// ============================================
// Main Entry Point
// ============================================

async function main() {
  try {
    console.log('🚀 Initializing DHT11 Sensor Control Agent...');
    console.log(`🌍 EC2 Sensor Control API: ${SENSOR_CONTROL_API_URL}`);
    console.log(`📍 Local IP: ${getLocalIP() || 'unknown'}`);
    console.log('ℹ️  Mode: control-only (no sensor data upload)');

    await registerSensor();

    const server = app.listen(HTTP_PORT, '0.0.0.0', () => {
      console.log(`🌐 Local control server started on port ${HTTP_PORT}`);
      console.log('Endpoints: /sensor/status, /sensor/control?action=on|off, /health');
    });

    process.on('SIGINT', () => {
      console.log('\n🛑 Shutting down gracefully...');
      server.close(() => {
        console.log('✅ Goodbye!');
        process.exit(0);
      });
    });

    statusMonitorLoop().catch((error) => {
      console.error(`❌ Fatal error in status monitor: ${error.message}`);
      process.exit(1);
    });
  } catch (error) {
    console.error(`❌ Fatal error: ${error.message}`);
    process.exit(1);
  }
}

main();
